---
layout: post
title: "Bioinformatics vs Computational Biology"
date:   2018-11-24 21:55:35 +0000
categories: rblogging
tags: [bioinformatics, computational, biology, word2vec]
---
A couple of weeks ago I saw this question on Twitter:

<blockquote class="twitter-tweet tw-align-center" data-theme="dark" data-link-color="#2B7BB9">
<p lang="en" dir="ltr">
Here's a question:<br><br>Do you think that “bioinformatics” and
“computational biology” are interchangeable?<br><br>Feel free to comment
to explain your answer.
</p>
— Biotweeps - Landon Getz (@biotweeps)
<a href="https://twitter.com/biotweeps/status/1058241544031940608?ref_src=twsrc%5Etfw">November
2, 2018</a>
</blockquote>
<script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
If we see the poll results and the replies to this tweet, we see that
people agree that they’re different, they just don’t know how.

This got me thinking that the answer probably lies on the science that
has been published so far, and if we could map the topics associated
with these fields we could have a (possible) answer. From [Jean Fan’s
blog](https://jef.works/blog/2018/02/06/fun-with-word2vec/) I had the
idea of using the same Word2Vec model of PubMed. Word2Vec in a neural
network method that produces a multidimensional embedding of the words
from a corpus of text. Each word is represented by a numeric vector, and
so we can perform vector operations on these terms.

Let’s start by downloading the [PubMed pre-trained model](http://bio.nlplab.org/) (approx. 3Gb):

``` bash
wget http://evexdb.org/pmresources/vec-space-models/PubMed-w2v.bin
```

Load the required libraries for the analysis. We will be using the
[wordVectors package](https://github.com/bmschmidt/wordVectors) to work
with the model

``` r
library(wordVectors) # manipulate word2vec things
library(tidyverse) # manipulate data and things
library(umap) # projection of things
library(RANN) # nearest neighbours of things
library(igraph) # graphs of things
library(ggrepel) # repels things

theme_set(theme_classic()+theme(aspect.ratio = 1))
```

And load the model (5-10min). This becomes hard to handle even on a
16Gb machine, so I’m only loading the 1450000 more frequent terms.

``` r
model = read.binary.vectors("PubMed-w2v.bin", nrows = 1450000)
gc(verbose = F) # clearing up some RAM
```

We can now start looking at word similarities. Because we’ll be using
this as our word set for further analysis, I will also throw in a few
other terms

``` r
similarities = list(
bioinf <- model %>% closest_to("bioinformatics", n=1000),

compbio <- model %>% closest_to(~"computational"+"biology", n=1000), # shared between computational and biology

comp <- model %>% closest_to(~"computational", n=1000),

compbio_only <- model %>% closest_to(~"computational"+"biology"-"bioinformatics", n=1000), # CB but not bioinf

comp_only <- model %>% closest_to(~"computational"-"bioinformatics", n=1000),

# bioinf_only <- model %>% closest_to(~"bioinformatics"-("computational"+"biology"), n=1000), # not run, outliers only

bioinf_only_comp <- model %>% closest_to(~"bioinformatics"-"computational", n=1000),

genomics <- model %>% closest_to("genomics", n=1000),

bio <- model %>% closest_to("biology", n=1000),

datasci <- model %>% closest_to(~"data"+"science"+"statistics", n=1000), # more accurate

strucbi <- model %>% closest_to(~"structure"+"biology", n=1000), # using "structural" gives weird results

sysbi <- model %>% closest_to(~"systems"+"biology", n=1000),

physics <- model %>% closest_to("physics", n=1000),

bio_physics <- model %>% closest_to(~"biology"+"physics", n=1000), # similar to "biophysics"

bio_math <- model %>% closest_to(~"biology"+"mathematics", n=1000) # not that similar to "biomathematics"
) %>% reduce(full_join, by = "word")
gc(verbose = F)

# only including words with a similarity>0.5 with at least one term
sim = apply(similarities[,2:14],1,function(x) any(x>=0.5, na.rm = T)) 
similarities = similarities[sim,]
true_labs = colnames(similarities)
colnames(similarities) = make.names(colnames(similarities)) # ugly valid names
```

Let’s have a first look at bioinformatics and computational biology
similarities

``` r
names_plot = rbind(top_n(similarities, 25, similarity.to..bioinformatics.), 
                   top_n(similarities, 25, similarity.to..computational.....biology.)) # these appear in both

ggplot(similarities, aes(x = similarity.to..bioinformatics.,
                         y = similarity.to..computational.....biology.))+
  geom_point(colour = "grey40")+
  geom_text(data = names_plot, mapping = aes(label = word), 
            fontface = "bold", size = 2.7)+
  geom_abline(slope = 1, intercept = 0)+
  labs(x = true_labs[2], y = true_labs[3])
```

<img src="bioinf_files/figure-markdown_github/unnamed-chunk-5-1.png" style="display: block; margin: auto;" />

It appears that in general there are more “-omics” terms towards
bioinformatics. Computational Biology has terms like modeling, genetics
and bioengineering. Lets now take a look at the difference between
computational biology and bioinformatics

``` r
arrange(similarities, desc(similarity.to..computational.....biology.....bioinformatics.))[1:25,] %>%
  ggplot(aes(x = reorder(word, desc(similarity.to..computational.....biology.....bioinformatics.)), 
             y = similarity.to..computational.....biology.....bioinformatics.))+
  geom_bar(stat = "identity")+
  labs(x = "Top 25 terms", y = true_labs[5])+
  theme(axis.text.x = element_text(angle = 30, hjust = 1),
        aspect.ratio = 1/1.5)
```

<img src="bioinf_files/figure-markdown_github/unnamed-chunk-6-1.png" style="display: block; margin: auto;" />

There are more (fascinating) physics terms. Maybe this is because the
word “computational” will be closer to physics terms. We can use UMAP on
the word vectors for our selected similar terms to see how they are
structured (and we’ll colour by similarity to bioinformatics)

``` r
model_subset = model[similarities$word,]
set.seed(1)
custom.config = umap::umap.defaults
custom.config$metric = "cosine"
umap_proj = umap::umap(model_subset@.Data, custom.config)
colnames(umap_proj$layout) = c("dim1", "dim2")

plot_df = merge(similarities, umap_proj$layout, by.x = 1, by.y = 0)
plot_df %>% 
  arrange(!is.na(similarity.to..bioinformatics.),
          similarity.to..bioinformatics.) %>% 
  ggplot(aes(x = dim1, y = dim2, colour = similarity.to..bioinformatics.))+
  geom_point()+
  labs(colour = true_labs[2])+
  guides(colour = guide_colourbar(title.position = "top"))+
  theme(legend.position = "bottom")
```

<img src="bioinf_files/figure-markdown_github/unnamed-chunk-7-1.png" style="display: block; margin: auto;" />

We can also cluster the data

``` r
louvainCluster = function(dat, knn = 30){
  # get knn
  knn.info = nn2(dat, k = knn)
  
  # transform into adjacency matrix
  knn = knn.info$nn.idx
  adj = matrix(0, nrow(model_subset), nrow(model_subset))
  rownames(adj) = colnames(adj) = rownames(model_subset)
  for(i in seq_len(nrow(model_subset))) {
      adj[i,rownames(model_subset)[knn[i,]]] = 1
  }
  
  # adjacency matrix to graph and cluster
  g = graph.adjacency(adj, mode = "undirected")
  g = simplify(g)
  
  km = cluster_louvain(g)
  com = paste0("cl", km$membership)
  names(com) <- km$names
  
  return(com)
}

clusts = louvainCluster(model_subset, knn = 50)
plot_df$cluster = clusts[plot_df$word]
ggplot(plot_df, aes(x = dim1, y = dim2, colour = cluster))+
  geom_point()
```

<img src="bioinf_files/figure-markdown_github/unnamed-chunk-8-1.png" style="display: block; margin: auto;" />

Interesting how there is a side cluster for terms related to
bioinformatics. We can have a better perspective by colouring the UMAP
by the maximum similarity of a term to each specific field

``` r
m = apply(plot_df[,c(2,3,8:15)], 1, function(x) true_labs[c(2,3,8:15)][which.max(x)])
m[is.na(lapply(m, nchar)==0)] = NA
plot_df$m = unlist(substr(m, 12, 100))
plot_df$m[plot_df$m==""] = NA

labels_df = plot_df[plot_df$word %in% c("bioinformatics", "computational", "biology", "genomics", "modeling",
                                        "epigenetics", "biochemistry", "dynamics", "structure",
                                        "biophysics", "mathematics", "biomedicine", "transcriptome", "BLAST",
                                        "superconductors", "Biostatistical", "optics", "acoustics"),]
plot_df %>% 
  arrange(!is.na(m)) %>%
  ggplot(aes(x = dim1, y = dim2, colour = m, label = word))+
  geom_point()+
  ggrepel::geom_text_repel(data = labels_df, colour = "black", min.segment.length = unit(0, "cm"),
                           mapping = aes(label = word))+ #ADD SOME WORDS
  labs(colour = "Max. simililarity")
```

<img src="bioinf_files/figure-markdown_github/unnamed-chunk-9-1.png" style="display: block; margin: auto;" />

This projection seems to imply that bioinformatics is heavily associated
with -omics, but also with databases and algorithms (present in the same
region as BLAST). The term “computational” is grouped closer to
“modeling” and physics related terms. While this is no decisive evidence
that computational biology has a closer connection to biophysics, it’s
worth noting that PubMed collects biomedical literature, hence that will
be the context of all these terms. It is also interesting to see the
association of strutural biology with some physics terms, and that
“biology” has “mathematics” and “Biostatistical” on the way to
“bioinformatics” (although there are quite a few weird terms in the
light blue region).

You can find the interactive plot below to explore yourself

``` r
plt = plot_df %>% 
  arrange(!is.na(m)) %>%
  ggplot(aes(x = dim1, y = dim2, colour = m, label = word))+
  geom_point()+
  labs(colour = "Max. simililarity")

p = plotly::ggplotly(plt)
# don't forget to set up an account to publish!
Sys.setenv("plotly_username" = "myuser")
Sys.setenv("plotly_api_key" = "mykey")
f = plotly::api_create(p)
```

<iframe width="785" height="600" frameborder="0" scrolling="no" src="//plot.ly/~tomasgomes/1.embed"></iframe>

<br />
There are also some limitations to this analysis. We are not truly
looking at “computational biology”, just assuming that these two words
will make the context something similar. Another thing is the presence
of some repeated words because of different cases of singular plural.
While the model isn’t perfect, if definitely provides an interesting
view on the relationship of these fields in the literature, showing how
scientists themselves prefer to use them.
